// @pm2/io/build/main/src/profiling/profilingCPUFallback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const debug_1 = require("debug");
const debug = debug_1.default('axm:profiling');
const file_1 = require("../utils/file");
const module_1 = require("../utils/module");
const configuration_1 = require("../configuration");
class ProfilingCPUFallback {
    constructor() {
        this.nsCpuProfiling = 'km-cpu-profiling';
        this.MODULE_NAME = 'v8-profiler-node8';
        this.FALLBACK_MODULE_NAME = 'v8-profiler';
    }
    init() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let path;
            let moduleName = this.MODULE_NAME;
            try {
                path = yield module_1.default.getModulePath(this.MODULE_NAME);
            }
            catch (e) {
                try {
                    moduleName = this.FALLBACK_MODULE_NAME;
                    path = yield module_1.default.getModulePath(this.FALLBACK_MODULE_NAME);
                }
                catch (err) {
                    configuration_1.default.configureModule({
                        heapdump: false
                    });
                    throw new Error('Profiler not loaded !');
                }
            }
            this.profiler = module_1.default.loadModule(path, moduleName);
            const enable = !(this.profiler instanceof Error);
            configuration_1.default.configureModule({
                heapdump: enable
            });
        });
    }
    destroy() {
        debug('Profiler destroyed !');
    }
    start() {
        this.profiler.startProfiling(this.nsCpuProfiling);
    }
    stop() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.getProfileInfo();
        });
    }
    getProfileInfo() {
        const cpu = this.profiler.stopProfiling(this.nsCpuProfiling);
        return file_1.default.writeDumpFile(JSON.stringify(cpu));
    }
}
exports.default = ProfilingCPUFallback;
