// @pm2/io/build/main/src/features/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const domain = require("domain");
const debug_1 = require("debug");
const debug = debug_1.default('axm:actions');
const serviceManager_1 = require("../serviceManager");
const transport_1 = require("../utils/transport");
const actions_1 = require("../services/actions");
class ActionsFeature {
    constructor(autoExit) {
        serviceManager_1.ServiceManager.set('actionsService', new actions_1.default(this));
        this.actionsService = serviceManager_1.ServiceManager.get('actionsService');
        process.on('message', this.listener);
        if (autoExit) {
            // clean listener if event loop is empty
            // important to ensure apm will not prevent application to stop
            this.timer = setInterval(() => {
                const dump = serviceManager_1.ServiceManager.get('eventLoopService').inspector.dump();
                if (!dump || (dump.setImmediates.length === 0 &&
                    dump.nextTicks.length === 0 &&
                    (Object.keys(dump.handles).length === 0 || (Object.keys(dump.handles).length === 1 &&
                        dump.handles.hasOwnProperty('Socket') &&
                        dump.handles.Socket.length === 2 &&
                        (dump.handles.Socket[0].fd === 1 ||
                            dump.handles.Socket[0].fd === -1) &&
                        (dump.handles.Socket[1].fd === 2 ||
                            dump.handles.Socket[1].fd === -1))) &&
                    Object.keys(dump.requests).length === 0)) {
                    process.removeListener('message', this.listener);
                }
            }, 1000);
            this.timer.unref();
        }
    }
    listener(data) {
        if (!data)
            return false;
        const actionName = data.msg ? data.msg : data.action_name ? data.action_name : data;
        let actionData = serviceManager_1.ServiceManager.get('actions').get(actionName);
        let fn = actionData ? actionData.fn : null;
        const reply = actionData ? actionData.reply : null;
        if (actionData) {
            // In case 2 arguments has been set but no options has been transmitted
            if (fn.length === 2 && typeof (data) === 'string' && data === actionName) {
                return fn({}, reply);
            }
            // In case 1 arguments has been set but options has been transmitted
            if (fn.length === 1 && typeof (data) === 'object' && data.msg === actionName) {
                return fn(reply);
            }
            /**
             * Classical call
             */
            if (typeof (data) === 'string' && data === actionName) {
                return fn(reply);
            }
            /**
             * If data is an object == v2 protocol
             * Pass the opts as first argument
             */
            if (typeof (data) === 'object' && data.msg === actionName) {
                return fn(data.opts, reply);
            }
        }
        // -----------------------------------------------------------
        //                      Scoped actions
        // -----------------------------------------------------------
        if (data.uuid === undefined || data.action_name === undefined) {
            return false;
        }
        actionData = serviceManager_1.ServiceManager.get('actionsScoped').get(actionName);
        if (data.action_name === actionName) {
            const res = {
                send: (dt) => {
                    transport_1.default.send({
                        type: 'axm:scoped_action:stream',
                        data: {
                            data: dt,
                            uuid: data.uuid,
                            action_name: actionName
                        }
                    });
                },
                error: (dt) => {
                    transport_1.default.send({
                        type: 'axm:scoped_action:error',
                        data: {
                            data: dt,
                            uuid: data.uuid,
                            action_name: actionName
                        }
                    });
                },
                end: (dt) => {
                    transport_1.default.send({
                        type: 'axm:scoped_action:end',
                        data: {
                            data: dt,
                            uuid: data.uuid,
                            action_name: actionName
                        }
                    });
                }
            };
            const d = domain.create();
            d.on('error', function (err) {
                res.error(err.message || err.stack || err);
                setTimeout(function () {
                    process.exit(1);
                }, 300);
            });
            d.run(function () {
                actionData.fn(data.opts || null, res);
            });
        }
    }
    init(conf, force) {
        this.actionsService.init(conf, force);
        return {
            action: this.action
        };
    }
    destroy() {
        this.actionsService.destroy();
        serviceManager_1.ServiceManager.reset('actions');
        serviceManager_1.ServiceManager.reset('actionsScoped');
        process.removeListener('message', this.listener);
        clearInterval(this.timer);
    }
    action(actionName, opts, fn) {
        if (!fn) {
            fn = opts;
            opts = null;
        }
        const check = this.check(actionName, fn);
        if (!check) {
            return check;
        }
        let type = 'custom';
        if (actionName.indexOf('km:') === 0 || actionName.indexOf('internal:') === 0) {
            type = 'internal';
        }
        // Notify the action
        transport_1.default.send({
            type: 'axm:action',
            data: {
                action_name: actionName,
                action_type: type,
                opts: opts,
                arity: fn.length
            }
        });
        const reply = (data) => {
            transport_1.default.send({
                type: 'axm:reply',
                data: {
                    return: data,
                    action_name: actionName
                }
            });
        };
        serviceManager_1.ServiceManager.get('actions').set(actionName, { fn: fn, reply: reply });
    }
    scopedAction(actionName, fn) {
        const check = this.check(actionName, fn);
        if (!check) {
            return check;
        }
        // Notify the action
        transport_1.default.send({
            type: 'axm:action',
            data: {
                action_name: actionName,
                action_type: 'scoped'
            }
        });
        serviceManager_1.ServiceManager.get('actionsScoped').set(actionName, { fn: fn });
    }
    check(actionName, fn) {
        if (!actionName) {
            console.error('[PMX] action.action_name is missing');
            return false;
        }
        if (!fn) {
            console.error('[PMX] callback is missing');
            return false;
        }
        if (!process.send) {
            debug('Process not running within PM2');
            return false;
        }
        return true;
    }
}
exports.default = ActionsFeature;
