// @pm2/io/build/main/src/features/coverage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const serviceManager_1 = require("../serviceManager");
const file_1 = require("../utils/file");
const semver = require("semver");
const debug_1 = require("debug");
const debug = debug_1.default('axm:coveragefeature');
class CoverageFeature {
    constructor() {
        this.inspectorService = serviceManager_1.ServiceManager.get('inspector');
    }
    init() {
        if (semver.satisfies(process.version, '< 8.0.0') ||
            (semver.satisfies(process.version, '< 10.0.0') && !process.env.FORCE_INSPECTOR)) {
            return new Promise(resolve => {
                debug(`Coverage feature is not available for node < 8.0.0 (force inspector : ${process.env.FORCE_INSPECTOR}), current version ${process.version}`);
                resolve(`Coverage feature is not available for node < 8.0.0 (force inspector : ${process.env.FORCE_INSPECTOR}), current version ${process.version}`);
            });
        }
        this.inspectorService.createSession();
        this.inspectorService.connect();
        return this.inspectorService.post('Profiler.enable');
    }
    destroy() {
        this.inspectorService.disconnect();
    }
    start(opts) {
        this.method = !opts.method ? 'takePreciseCoverage' : opts.method;
        return this.inspectorService.post('Profiler.startPreciseCoverage', { callCount: !!opts.callCount, detailed: !!opts.detailed });
    }
    stop() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const data = yield this.getInfo();
            return file_1.default.writeDumpFile(JSON.stringify(data), '.coverageprofile');
        });
    }
    getInfo() {
        // const self = this
        return new Promise((resolve, reject) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                const data = yield this.inspectorService.post('Profiler.' + this.method);
                yield this.inspectorService.post('Profiler.stopPreciseCoverage');
                return resolve(data);
            }
            catch (err) {
                debug('Coverage profiling stopped !', err);
                return reject(err);
            }
        }));
    }
}
exports.default = CoverageFeature;
